server/src/main/java/org/opensearch/cluster/service/MasterTaskThrottler.java

```diff
@@ -0,0 +1,139 @@
+import org.opensearch.action.admin.cluster.snapshots.delete.DeleteSnapshotRequest;
+import org.opensearch.action.admin.indices.create.CreateIndexRequest;
+import org.opensearch.action.admin.indices.dangling.delete.DeleteDanglingIndexRequest;
+import org.opensearch.action.admin.indices.mapping.put.PutMappingClusterStateUpdateRequest;
+import org.opensearch.common.metrics.CounterMetric;
+import org.opensearch.common.settings.ClusterSettings;
+import org.opensearch.common.settings.Setting;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.snapshots.UpdateIndexShardSnapshotStatusRequest;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * This class is extension of {@link Throttler} and does throttling of master tasks.
+ *
+ * This class does throttling on task submission to master, it uses class name of request of tasks as key for
+ * throttling. Throttling will be performed over task class level, different task types have different request class, i.e.
+ * Throttling is at the task type level.
+ *
+ * Set "master.throttling.enable" setting to enable/disable this throttling. (Default False)
+ * Set specific setting to for setting the threshold of throttling of particular task type.
+ * e.g : Set "master.throttling.thresholds.put_mapping" to set throttling limit of "put mapping" tasks,
+ *       Set it to default value(-1) to disable the throttling for this task type.
+ *
+ *       We can also provide the class path of request, which we want to throttle. If new task type has introduced and we
+ *       haven't added that in configuration then we can use the class path of request to enable throttling on that type.
+ *       e.g use "org/opensearch/action/admin/indices/create/CreateIndexClusterStateUpdateRequest" as key in setting
+ *       throttling limit for Create index tasks.
+ */
+public class MasterTaskThrottler extends Throttler<Class> {
+    private static final Logger logger = LogManager.getLogger(MasterTaskThrottler.class);
+
+    public static final Setting<Boolean> ENABLE_MASTER_THROTTLING =
+            Setting.boolSetting("master.throttling.enable", false,
+                    Setting.Property.Dynamic, Setting.Property.NodeScope);
+
+    public static final Setting<Settings> THRESHOLD_SETTINGS =
+            Setting.groupSetting("master.throttling.thresholds.", Setting.Property.Dynamic, Setting.Property.NodeScope);
+
+    protected final ConcurrentHashMap<String, Class> taskNameToClassMapping = new ConcurrentHashMap() {
+        {
+            put("put_mapping", PutMappingClusterStateUpdateRequest.class);
+            put("create_index", CreateIndexRequest.class);
+            put("delete_dangling_index", DeleteDanglingIndexRequest.class);
+            put("update_snapshot", UpdateIndexShardSnapshotStatusRequest.class);
+            put("delete_snapshot", DeleteSnapshotRequest.class);
+        }
+    };
+    private final int DEFAULT_THRESHOLD_VALUE = -1; // Disabled throttling
```

Comment:
For minor version upgrades same scenario mentioned by you can cause issue (New master / old data nodes) If customer changes the threshold limit during upgrade and it starts throwing new exception. In that case it would be breaking change. 

Also we are introducing new settings, will it also be breaking change as older minor version nodes wont be able to understand new setting if customer has set it during upgrade.

> Ideally it is expected for data nodes to upgrade first and then master nodes and then switch the active master as data nodes in lower version may not be able to connect to master in high version

Does it stand for minor version upgrade as well? 
If yes, then I think we are covered in above two cases for minor version upgrades and expecting customer to follow this process for upgrade.

Reply:
> For minor version upgrades same scenario mentioned by you can cause issue (New master / old data nodes) If customer changes the threshold limit during upgrade and it starts throwing new exception. In that case it would be breaking change.
> 
We can recommend customers to not enable this setting during upgrade otherwise their request could throttle without any automatic retries. They should enable this setting once the upgrade has finished. 

>Also we are introducing new settings, will it also be breaking change as older minor version nodes wont be able to understand new setting if customer has set it during upgrade.
>
Addition of new settings shouldn't be a breaking change, even on old nodes, can you try it?

>Does it stand for minor version upgrade as well?
>
Ideally, it applies to all version upgrades minor or major.

Reply:
> We can recommend customers to not enable this setting during upgrade otherwise their request could throttle without any automatic retries. They should enable this setting once the upgrade has finished.

For that I have added validation of min version in cluster while updating setting. I have added that validation in enable/disable setting. Will move it to threshold setting as we are relying on it only. 

> Addition of new settings shouldn't be a breaking change, even on old nodes, can you try it?

Actually, I had tried it. If new setting is applied and we call _cluster/settings API from older version node then, it would fail to parse new setting. We are recommending not to enable this setting during upgrade so I think that will cover this case as well. 
What do you think? 

> Ideally, it applies to all version upgrades minor or major.

Okay, yes it make sense to follow this for all upgrades.